### PR TITLE
Code gen for tuple instances

### PR DESCRIPTION
**Pull request by adelbertc**

### CODE REVIEW

**Comment by non** [comment]
This looks great.

Could you add support for (at least) `PartialOrder`, `Semilattice`, `EuclideanRing`, and `Field`?

In theory we could support all the other variants (e.g. the commutative versions, `Bands`, bounded lattices) but I'm willing to wait and see if anyone asks for those.

File: laws/shared/src/test/scala/algebra/laws/LawTests.scala [hunk 1]
```diff
@@ -89,13 +89,7 @@ trait LawTestsBase extends FunSuite with Discipline {
   laws[RingLaws, BigInt].check(_.euclideanRing)
 
   {
-    implicit val tupEq: Eq[(Int, Int)] = new Eq[(Int, Int)] {
-      def eqv(a: (Int, Int), b: (Int, Int)) = a == b
-    }
-    implicit val band = new Band[(Int, Int)] {
-      def combine(a: (Int, Int), b: (Int, Int)) = (a._1, b._2)
-    }
-    checkAll("(Int, Int) Band", GroupLaws[(Int, Int)].band)
```
Comment (johnynek) [inline laws/shared/src/test/scala/algebra/laws/LawTests.scala]: why did you delete this band test? If nothing else it was exercising GroupLaws.band.

File: project/Boilerplate.scala [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        -      def mod(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("mod")}
+        -      def negate(x: ${`(A..N)`}): ${`(A..N)`} = ${unaryTuple("negate")}
+        -      def one: ${`(A..N)`} = ${nullaryTuple("one")}
+        -      def plus(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("plus")}
+        -      def quot(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("quot")}
+        -      def times(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("times")}
+        -      def zero: ${`(A..N)`} = ${nullaryTuple("zero")}
+        -    }
+        -
+        -  implicit def tuple${arity}Field[${`A..N`}](implicit ${constraints("Field")}): Field[${`(A..N)`}] =
```
Comment (johnynek) [inline project/Boilerplate.scala]: fields don't compose in this way. It won't pass the laws (there e.g. (0, 1), (1, 0) are neither zero nor do they have inverses).

File: project/Boilerplate.scala [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        -      def quot(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("quot")}
+        -      def times(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("times")}
+        -      def zero: ${`(A..N)`} = ${nullaryTuple("zero")}
+        -    }
+        -
+        -  implicit def tuple${arity}Monoid[${`A..N`}](implicit ${constraints("Monoid")}): Monoid[${`(A..N)`}] =
+        -    new Monoid[${`(A..N)`}] {
+        -      def combine(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("combine")}
+        -      def empty: ${`(A..N)`} = ${nullaryTuple("empty")}
+        -    }
```
Comment (johnynek) [inline project/Boilerplate.scala]: ideally we would compose `combineAll` in some way. At least add an issue. For many monoids the naive combineAll is very slow (for strings and lists it is `O(N^2)`, for Maps using an intermediate mutable map is much faster).

Comment (johnynek) [inline project/Boilerplate.scala]: same is true for semigroup, of course.

File: project/Boilerplate.scala [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        -      def zero: ${`(A..N)`} = ${nullaryTuple("zero")}
+        -      def negate(x: ${`(A..N)`}): ${`(A..N)`} = ${unaryTuple("negate")}
+        -    }
+        -
+        -  implicit def tuple${arity}Semigroup[${`A..N`}](implicit ${constraints("Semigroup")}): Semigroup[${`(A..N)`}] =
+        -    new Semigroup[${`(A..N)`}] {
+        -      def combine(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("combine")}
+        -    }
+        -
+        -  implicit def tuple${arity}Semilattice[${`A..N`}](implicit ${constraints("Semilattice")}): Semilattice[${`(A..N)`}] =
```
Comment (johnynek) [inline project/Boilerplate.scala]: any reason to leave Band out?

**Comment by johnynek** [comment]
@non you steered him wrong on Field, my friend.

**Comment by johnynek** [comment]
Of course, it would be cool to have a Galois fields for tuples of Boolean. :) (which is the one exception that pops to mind for standard scala types, but it does not compose this way).

**Comment by adelbertc** [comment]
@johnynek Thanks for your feedback, will address them soon! I'm out of town for the weekend so most likely won't be able to do anything until next week.

File: project/Boilerplate.scala (hunk 2 above)
Comment (non) [inline project/Boilerplate.scala]: I think I suggested that @adelbertc add this, but yes, we should remove it. Sorry!

File: project/Boilerplate.scala (hunk 3 above)
Comment (non) [inline project/Boilerplate.scala]: +1 on these. Let's open an issue for now.

File: project/Boilerplate.scala [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        -      def eqv(x: ${`(A..N)`}, y: ${`(A..N)`}): Boolean = ${binMethod("eqv").mkString(" && ")}
+        -    }
+        -
+        -  implicit def tuple${arity}EuclideanRing[${`A..N`}](implicit ${constraints("EuclideanRing")}): EuclideanRing[${`(A..N)`}] =
+        -    new EuclideanRing[${`(A..N)`}] {
+        -      def mod(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("mod")}
+        -      def negate(x: ${`(A..N)`}): ${`(A..N)`} = ${unaryTuple("negate")}
+        -      def one: ${`(A..N)`} = ${nullaryTuple("one")}
+        -      def plus(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("plus")}
+        -      def quot(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("quot")}
```
Comment (johnynek) [inline project/Boilerplate.scala]: I'm skeptical of this one as well. Can you add a test that checks that `(Int, Int)` forms a EuclideanRing? I don't know as much about the theory of euclidean rings, so I don't know off hand the answer.

**Comment by non** [comment]
:+1:

**Comment by non** [comment]
@johnynek I encouraged @adelbertc to add the euclidean ring instance (as well as the erroneous field instance). I'm pretty sure it's OK (although I haven't written a proof yet) but if you are nervous I would be willing to remove it (and obviously apologize to @adelbertc for suggesting the work).

**Comment by johnynek** [comment]
Yeah, I guess it works for a Euclidean Ring, but I noticed that we don't actually have a test for Euclidean Rings. It is currently a TODO and actually just checking the Ring laws.

+1 from me. Added #85
